### PR TITLE
main/python3: upgrade to 3.6.2, optimize build, cleanup

### PR DESCRIPTION
**Pull request by scadu**
Spitted into two commits since https://github.com/scadu/aports/commit/dcf6247dc0426b59bcbce8deefcd852dc7e02820 which fixes issue reported by @awilfox on IRC:

> I am building packages in a chroot.  the host system is 64-bit Debian, the chroot is 32-bit Alpine, but it says: checking host system type... powerpc64-unknown-linux-gnu
this is definitely not desireable.  it should be using --build --host like any other

The second commit https://github.com/scadu/aports/commit/6c482bda93672107fa892d2a1ed213da31ce2b39 actually enables optimization options somehow missed during previous builds. 

Travis test will fail here. Unfortunately it requires more time, however the boost is noticeably: 30% speed boost in some parts of the stdlib -- I think it's worth it.

### CODE REVIEW

**Comment by awilfox** [comment]
I can confirm this solves our chroot build/host type issue.

The optimisations are indeed a very good idea, but see docker-library/python#160.  When I applied this to our aforementioned PowerPC builder, I had some very noteworthy results:

1. The build process took over two hours.  Of course, it is a dual-processor 1.8 GHz Power4, so it is not exactly a shiny new build cluster.
2. When `abuild` hit the package phase, `make clean` was run by Python's build system, and the entire thing started over again.  I started the build at 20:58 and by 01:15, it was just wrapping up.
3. Unfortunately while making the apk files, the chroot ran out of disk space.  No worries, I thought; just increase the size of the volume and run `abuild rootpkg` again.  Of course, that started the build over again, since Python did `make clean` again...
4. And then this happened:

```
/tmp/ccObloFK.ltrans4.ltrans.o: In function `_GLOBAL__sub_I_65535_0_pgen.c':
<artificial>:(.text.startup+0x24): undefined reference to `__gcov_init'
/usr/lib/gcc/powerpc-foxkit-linux-musl/6.3.0/../../../../powerpc-foxkit-linux-musl/bin/ld: BFD (GNU Binutils) 2.28 assertion fail elf32-ppc.c:8923
/usr/lib/gcc/powerpc-foxkit-linux-musl/6.3.0/../../../../powerpc-foxkit-linux-musl/bin/ld: BFD (GNU Binutils) 2.28 assertion fail elf32-ppc.c:8923
/tmp/ccObloFK.ltrans5.ltrans.o:/home/awilcox/aports/main/python3/src/Python-3.6.1/Parser/pgen.c:254: more undefined references follow
collect2: error: ld returned 1 exit status
make: *** [Makefile:813: Parser/pgen] Error 1
>>> ERROR: python3*: package failed
>>> ERROR: python3: rootpkg failed
```

At this point I was having major Gentoo flashbacks and gave up.  :wink:  Maybe --enable-optimizations is not stable on PowerPC or maybe it is incompatible with GCC 6.3.0.  Not sure how to tell, and not sure I want to spend the machine cycles and wall time on debugging it at this time.  But do feel free to try and figure out, if you want, because it is a good thing to do.

**Comment by scadu** [comment]
The first three "problems" are the result of enabled flags. That's how it works and it's explained in linked docker-library issue. 

If it comes to the last one: I'm afraid I have no access to any ppc machine, and even so debugging this would be a magic. We could omit those flags on ppc till we figure it out. Thanks.

**Comment by scadu** [comment]
Maybe we should apply ` --with-lto` and `--enable-optimizations` only for x86` and x86_64.
@jirutka, what do you think?

**Comment by awilfox** [comment]
LTO works fine on PowerPC.  Can't speak for any other architecture, bootstrapping takes a while.

File: main/python3/APKBUILD [hunk 1]
```diff
@@ -40,17 +40,17 @@ build() {
...
-		--enable-optimize \
-		--enable-lto \
+		--enable-optimizations \
+		--with-lto \
 		--with-computed-gotos \
 		--with-dbmliborder=gdbm:ndbm \
 		--with-system-expat \
 		--with-system-ffi \
 		--with-threads \
+		--with-ensurepip=install \
```
Comment (ncopa) [inline main/python3/APKBUILD]: Why so you enable ensurepip? see http://bugs.alpinelinux.org/issues/5381

Comment (scadu) [inline main/python3/APKBUILD]: Looks like I forgot to remove this line once I finished playing with APKBUILD.

**Comment by scadu** [comment]
3.6.2 build with `--with-lto` and `--enable-optimizations` took 26 minutes on Xeon E31220 with 16G RAM and HDD. @ncopa would it be possible to merge all of this?

File: main/python3/APKBUILD [hunk 2]
```diff
@@ -40,17 +40,17 @@ build() {
 		--build=$CBUILD \
 		--host=$CHOST \
 		--prefix=/usr \
-		--disable-rpath \
```
Comment (ncopa) [inline main/python3/APKBUILD]: why was `--disable-rpath` removed?

File: main/python3/APKBUILD (hunk 1 above)
Comment (ncopa) [inline main/python3/APKBUILD]: i think the entire commit can be squashed with the previous

File: main/python3/APKBUILD [hunk 3]
```diff
@@ -50,7 +50,6 @@ build() {
 		--with-system-expat \
 		--with-system-ffi \
 		--with-threads \
-		--with-ensurepip=install \
 		
 
 	make EXTRA_CFLAGS="$CFLAGS"
```
Comment (ncopa) [inline main/python3/APKBUILD]: I think the commit should be squashed with the previous (`git rebase -i` and set this to `fixup`)

File: main/python3/APKBUILD [hunk 4]
```diff
@@ -2,7 +2,7 @@
 # Contributor: Kiyoshi Aman <kiyoshi.aman@gmail.com>
 
 pkgname=python3
-pkgver=3.6.1
+pkgver=3.6.2
 _basever="${pkgver%.*}"
 pkgrel=5
```
Comment (ncopa) [inline main/python3/APKBUILD]: we should set `pkgrel=0` when we bump `pkgver`

**Comment by scadu** [comment]
Following tests failed for me and @awilfox 

```
 test__locale failed
 test_asyncore skipped
 test_capi failed
 test_curses skipped (resource denied)
 test_devpoll skipped
 test_gdb skipped
 test_kqueue skipped
 test_locale failed
 test_msilib skipped
 test_multiprocessing_fork skipped
 test_multiprocessing_forkserver skipped
 test_multiprocessing_main_handling skipped
 test_multiprocessing_spawn skipped
 test_nis skipped
 test_os failed
 test_ossaudiodev skipped (resource denied)
 test_posix failed
 test_re failed
 test_resource failed
 test_shutil failed
 test_smtpnet skipped (resource denied)
 test_socketserver skipped (resource denied)
 test_startfile skipped
 test_strptime failed
 test_subprocess skipped
 test_threadsignals failed
 test_time failed
 test_timeout skipped (resource denied)
 test_tix skipped (resource denied)
 test_tk skipped (resource denied)
 test_ttk_guionly skipped (resource denied)
 test_urllib2net skipped (resource denied)
 test_urllibnet skipped (resource denied)
 test_winconsoleio skipped
 test_winreg skipped
 test_winsound skipped (resource denied)
 test_xmlrpc_net skipped (resource denied)
 test_zipfile64 skipped (resource denied)
```

Quoting @awilfox 

> test_locale is meant to fail; musl doesn't support strcoll (it is stub) nor strxfrm (which uses strcoll) nor nl_langinfo without you writing your own (which is why '.' != ',')

Probably the rest of tests failed due to musl too since it passes on glibc-based distro (e.g. Debian Jessie). There is no distro to compare since Void do not use optimization and they are still on 3.5.

File: main/python3/APKBUILD [hunk 5]
```diff
@@ -38,13 +38,13 @@ build() {
 	cd "$builddir"
 
 	./configure \
-		--prefix=/usr \
+		--prefix=/usr
```
Comment (mingwandroid) [inline main/python3/APKBUILD]: You probably don't want to remove the line continuation here.

Comment (scadu) [inline main/python3/APKBUILD]: Already fixed. Please see [204198c](https://github.com/alpinelinux/aports/pull/1775/commits/204198c974ccdc09154075cfe7d8eb617e73f85d#diff-6e86de19c7a7623f0f4c6030b8eb5940R38)

File: main/python3/APKBUILD (hunk 3 above)
Comment (scadu) [inline main/python3/APKBUILD]: Done.

File: main/python3/APKBUILD (hunk 4 above)
Comment (scadu) [inline main/python3/APKBUILD]: Done.

File: main/python3/APKBUILD (hunk 2 above)
Comment (scadu) [inline main/python3/APKBUILD]: Shouldn't be included in the commit. Restored.

**Comment by fabled** [comment]
I had to remove optimizations due to failure on x86 (probably also aarch64). See commit 3778f5bafc528d05f991945fb6dc5fa9781d1873.
